conan/tools/cmake/cmake.py

```diff
@@ -108,7 +108,7 @@ def _build(self, build_type=None, target=None, cli_args=None, build_tool_args=No
                                          "single-config build systems")
 
         bt = build_type or self._conanfile.settings.get_safe("build_type")
-        if not bt:
+        if is_multi and not bt:
```

Comment:
This protection for not defining the ``build_type`` was in place, necessary for CMake integration, because if ``build_type`` is not defined, it later fails in ``CMakeDeps`` with either a message like
> Please, set the CMAKE_BUILD_TYPE variable when calling to CMake "
                            "adding the '-DCMAKE_BUILD_TYPE=<build_type>' argument

Or directly missing dependencies.

Reply:
It is only that we get some error output in log, but the build succeeds. I will create some test for it to show you

Reply:
Sure, it will be good to have some tests. Please add consuming the package with ``test_package``, for example. 
Still the above feedback would be relevant: why going against CMake and not defining the build_type? What is the use case? Thanks!

Reply:
Note: `bt` is only used if `is_multi == True`. Otherwise it does not have any effect. This is why I have changed this here, since it would otherwise throw an exception for valid cases (see test)

Reply:
It has one effect: raising an exception. If ``bt`` is not defined it gives an error, saying it is required. Because it is, if we don't have ``build_type`` here, this package will fail later because it doesn't have ``build_type``, and ``CMakeDeps`` will fail. 

This exception is here to protect against a very common mistake that is forgetting the ``build_type`` when using ``CMake``, that CMake explicitly discourage not passing a ``build_type``. If we remove this protection, we will get again a lot of new tickets of users that were biten by this. If we want to remove this protective safety measure, it has to be clear what use case are we supporting.

Reply:
I completely agree that it should raise an exception in this case.
But there are also customers who are not using `CMakeDeps` at all, so for them it should not fail, or?

Why not handling the error in `CMakeDeps` where it causes the actual issues?

Reply:
> Still the above feedback would be relevant: why going against CMake and not defining the build_type? What is the use case?

Use case: We have some tool_requires with settings `"os", "compiler", "arch"`.
The setting `build_type` we want to ignore, so we want to always enforce to build `Release`.

We tried to specify `build_type` in different ways, but we face issues:

- If we use `cmake.build(build_type=build_type)` then conan logs error `ERROR: Don't specify 'build_type' at build time for single-config build systems`. In the code I can see, that this actually has no effect. Result: There is no `CMAKE_BUILD_TYPE` specified.

- Adding `build_type` to settings does not have any effect, since `cache_variables["CMAKE_BUILD_TYPE"]` is overwritten by settings. Result: We sometimes build with wrong `CMAKE_BUILD_TYPE=Debug`

Reply:
> The setting build_type we want to ignore, so we want to always enforce to build Release

Ok, this is very important. You don't want to build something without ``build_type``. You want to build it in ``Release`` build-type. This is a fresh perspective that doesn't goes against CMake recommendations, I'll have a new look with this in mind.

Reply:
There is a recommended flow, that we also implement in ConanCenter:
- Leave the ``build_type`` defined
  - This will make sure that everything that needs ``build_type`` will work. From ``CMakeDeps``, to the ``cmake_layout``
- Remove the ``build_type`` in the ``package_id()`` method, so usage of the package becomes independent from ``build_type``.
- Define ``build_type=Release`` in the profile to build this package, create it and upload it. From now all packages can ``tool_require`` it irrespective of the build_type

There is a further insight for this. Using the 2 profiles "build" and "host", it is no longer an issue or necessary to force or remove the ``build_type`` from ``tool_requires``, as the "build" context will almost always use the ``build_type=Release`` naturally. This is the most recommended approach, it is easy, built-in, doesn't require changes in recipe and is well tested and used in production by many users.

Reply:
Ok, found the gap.
You are right that it can make sense to not raise the exception for that case, and allow to force the ``Release`` in recipes.
However, it might make sense to have to define the recipe with explicit behavior like:
```python
    def layout(self):
        self.folders.build = "build"

    def generate(self):
        tc = CMakeToolchain(self)
        if self.settings.compiler != "msvc":
            tc.cache_variables["CMAKE_BUILD_TYPE"] = "Release"
        tc.generate()

    def build(self):
        cmake = CMake(self)
        cmake.configure()
        build_type = "Release" if self.settings.compiler == "msvc" else None
        cmake.build(build_type=build_type)

    def package(self):
        cmake = CMake(self)
        build_type = "Release" if self.settings.compiler == "msvc" else None
        cmake.install(build_type=build_type)
```
and then apply your change so it doesn't error. But not necessarily changing the ``error`` => ``warning`` message and still get the warning with would be still not great. It is better to not create the warning in the first place. That means that we might want also to make public some ``is_multi_configuration()`` interface, to simplify recipes and make them possible to do this.

I'll try to move this forward tomorrow, thanks for the feedback

Reply:
In any case, please read the comment above, all this would be unnecessary with the "build" profile defining ``build_type=Release``

Reply:
Thank you very much :-)

So summary from my POV: 
- it is very unintuitive that `cmake.build(build_type="Release")` does not have any effect. So it totally makes sense to warn users about this behavior in a WARN/ERR message.
- conan should raise exception if build type is not set and it is a multi configuration build
- conan should WARN/ERR users if build_type is empty
- In case of ` settings = "os", "compiler", "arch"` the `CMAKE_BUILD_TYPE` is empty, even if `build_type` setting is specified in the profile. conan could set `CMAKE_BUILD_TYPE=Release` by default (at least for single configuration --> proposal see this PR)